File: partitionEqualSubsetSum/partitionEqualSubsetSum.cpp
// clang++ partitionEqualSubsetSum.cpp -std=c++11

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int sum = 0;
        for (int num : nums) sum += num;
        if ((sum & 1) == 1) return false;
        sum /= 2;
        int size = nums.size();
        vector<bool> dp(sum + 1, false);
        dp[0] = true;
        for (int i = 1; i <= sum; i++)
            dp[i] = false;
        for (int i = 1; i <= size; i++) {
            for (int j = sum; j >= nums[i - 1]; j--) {
                if (j >= nums[i - 1]) {
                    dp[j] = dp[j] || dp[j - nums[i - 1]];
                }
            }
        }//for
        return dp[sum];
    }
};

int main()
{
    Solution s;
    vector<int> nums1 = {1,5,11,5};
    vector<int> nums2 = {1,2,3,5};
    cout << s.canPartition(nums1) << endl;
    cout << s.canPartition(nums2) << endl;
    return 0;
}

File: 416.partitionEqualSubsetSum/partitionEqualSubsetSum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int sum = 0;
        for (int num : nums) sum += num;
        if (sum % 2 != 0) return false;
        int target = sum / 2;
        boolean dp[][] = new boolean[nums.length + 1][target + 1];
        dp[0][0] = true;
        for (int i = 1; i <= nums.length; i++) {
            for (int j = 0; j <= target; j++) {
                if (dp[i - 1][j]) { 
                    dp[i][j] = true;
                } else if (j - nums[i - 1] >= 0) {
                    dp[i][j] = dp[i - 1][j - nums[i - 1]];
                }
            }
        }
        return dp[nums.length][target];
    }
}
